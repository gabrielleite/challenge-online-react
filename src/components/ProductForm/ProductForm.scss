h2 { margin: 0; margin-bottom: 30px; }

.input-wrapper {
  margin-bottom: 30px;

  .label {
    font-style: normal; font-weight: normal; font-size: 14px; line-height: 18px; 
    display: flex; align-items: center; color: #62595D; margin-bottom: 8px;
  }

  input {
    &[type="text"], &[type="number"] {
      border: 1px solid #BFBBBD;
      box-sizing: border-box; padding: 8px;
      border-radius: 8px; width: 100%;
    }
    font-size: 16px; line-height: 24px;
  }

  .validation-error-message {
    color: red;
  }
}

.button { 
  border: none; background: #2C3357;
  border-radius: 8px; padding: 8px 16px; min-width: 100px; font-weight: bold; color: #FFFFFF;
}
