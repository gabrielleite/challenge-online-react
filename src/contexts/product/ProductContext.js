import React, { createContext } from 'react';

export const ProductContext = createContext();

export const ProductContextProvider = ({children}) => {

    const add = (product) => {
        localStorage.setItem(new Date().getTime(), JSON.stringify(product));
    };

    const list = () => {
        return Object.keys(localStorage).map((key) => 
            JSON.parse(localStorage[key]));
    }

    return (
        <ProductContext.Provider value={ { 
            getProducts: list, 
            addProduct: add 
        } }>
            {children}
        </ProductContext.Provider>
    );
};