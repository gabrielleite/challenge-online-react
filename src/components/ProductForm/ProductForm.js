import React, { useContext } from 'react';
import { Button } from '@tiendanube/components';
import { DownloadIcon } from '@tiendanube/icons';
import { useForm } from 'react-hook-form';
import { ProductContext } from '../../contexts/product/ProductContext';

import './ProductForm.scss';

const ProductForm = () => {
  const { register, handleSubmit, formState: { errors }, reset } = useForm();
  const { addProduct } = useContext(ProductContext);

  const handleFormSubmit = (product) => {
    addProduct(product);
    reset();
  }

  return (
    <form className="centered" onSubmit={handleSubmit(handleFormSubmit)}>
      <h2>Add new product</h2>
      <div>
        <div className="input-wrapper">
          <label className="label" htmlFor="name">Product name</label>
          <input id="name" type="text" {...register('name', {required: true})} />
          {errors.name && <span className="validation-error-message">Product name is required</span>}
        </div>

        <div className="input-wrapper">
          <label className="label" htmlFor="count">Items count</label>
          <input id="count" type="number" defaultValue="0" min="0" {...register('count', { required: true })} />
          {errors.count && <span className="validation-error-message">Items count is required</span>}
        </div>

        <div className="input-wrapper">
          <label className="label" htmlFor="price">Price</label>
          <input id="price" type="number" min="0" step=".01" {...register('price', { required: true })} />
          {errors.price && <span className="validation-error-message">Price is required</span>}
        </div>

        <div className="input-wrapper">
          <label className="label" htmlFor="promotionalPrice">Promotional Price</label>
          <input id="promotionalPrice" type="number" min="0" step=".01" {...register('promotionalPrice', { required: true })} />
          {errors.promotionalPrice && <span className="validation-error-message">Promotional price is required</span>}
        </div>

        <Button appearance="primary" icon={DownloadIcon} type="submit">
          Aceptar
        </Button>
      </div>
    </form>
  );
}

export default ProductForm;
