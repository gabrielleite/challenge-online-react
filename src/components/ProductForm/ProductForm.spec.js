import '@testing-library/jest-dom/extend-expect';

import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import ProductForm from "./ProductForm";
import { ProductContext } from '../../contexts/product/ProductContext';

describe('ProductForm', () => {
    let wrapper;
    const productContextActions = {
        getProducts: jest.fn(),
        addProduct: jest.fn()
    };
    beforeEach(() => {
        wrapper = render(
            <ProductContext.Provider value={productContextActions}>
                <ProductForm />
            </ProductContext.Provider>
        );
    });

    it('renders correctly', () => {
        const { container } = wrapper;
        expect(container).toMatchSnapshot();
    });

    describe('accessibility interactions', () => {
        it('should find product name input by its label', () => {
            const inputElement = screen.getByLabelText('Product name');
            expect(inputElement).not.toBeUndefined();
        });

        it('should find product count input by its label', () => {
            const inputElement = screen.getByLabelText('Items count');
            expect(inputElement).not.toBeUndefined();
        });

        it('should find product price input by its label', () => {
            const inputElement = screen.getByLabelText('Price');
            expect(inputElement).not.toBeUndefined();
        });

        it('should find product promotionalPrice input by its label', () => {
            const inputElement = screen.getByLabelText('Promotional Price');
            expect(inputElement).not.toBeUndefined();
        });
    });

    describe('onSubmit', () => {
        describe('when invalid', () => {
            beforeEach(() => {
                const countField = screen.getByRole('spinbutton', { name: 'Items count' });
                const submitButton = screen.getByRole('button');
                userEvent.type(countField, ' ');

                userEvent.click(submitButton);
            });

            it('should show product name error message', () => {
                expect(screen.queryByText('Product name is required')).toBeInTheDocument();
                expect(productContextActions.addProduct).not.toHaveBeenCalled();
            });
    
            it('should show count error message', () => {
                expect(screen.queryByText('Items count is required')).toBeInTheDocument();
                expect(productContextActions.addProduct).not.toHaveBeenCalled();
            });

            it('should show product price error message', () => {
                expect(screen.queryByText('Price is required')).toBeInTheDocument();
                expect(productContextActions.addProduct).not.toHaveBeenCalled();
            });

            it('should show promotionalPrice error message', () => {
                expect(screen.queryByText('Promotional price is required')).toBeInTheDocument();
                expect(productContextActions.addProduct).not.toHaveBeenCalled();
            });
        });

        describe('when valid', () => {
            let productNameField;
            let countField;
            let priceField;
            let promotionalPriceField;

            beforeEach(() => {
                productNameField = screen.getByRole('textbox', { name: 'Product name' });
                countField = screen.getByRole('spinbutton', { name: 'Items count' });
                priceField = screen.getByRole('spinbutton', { name: 'Price' });
                promotionalPriceField = screen.getByRole('spinbutton', { name: 'Promotional Price' });
                const submitButton = screen.getByRole('button');

                userEvent.type(productNameField, 'Product1');
                userEvent.type(countField, '10');
                userEvent.type(priceField, '2,00');
                userEvent.type(promotionalPriceField, '1,50');

                userEvent.click(submitButton);
            });

            it('should add product', () => {
                expect(productContextActions.addProduct).toHaveBeenCalled();
            });

            it('should reset the form', () => {
                expect(productNameField.getAttribute('value')).toBeNull();
                expect(countField.getAttribute('value')).toBe('0');
                expect(priceField.getAttribute('value')).toBeNull();
                expect(promotionalPriceField.getAttribute('value')).toBeNull();
            });
        });
    });
});