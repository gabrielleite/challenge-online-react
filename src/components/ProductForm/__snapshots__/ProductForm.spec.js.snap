// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm renders correctly 1`] = `
<div>
  <form
    class="centered"
  >
    <h2>
      Add new product
    </h2>
    <div>
      <div
        class="input-wrapper"
      >
        <label
          class="label"
          for="name"
        >
          Product name
        </label>
        <input
          id="name"
          name="name"
          type="text"
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          class="label"
          for="count"
        >
          Items count
        </label>
        <input
          id="count"
          min="0"
          name="count"
          type="number"
          value="0"
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          class="label"
          for="price"
        >
          Price
        </label>
        <input
          id="price"
          min="0"
          name="price"
          step=".01"
          type="number"
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          class="label"
          for="promotionalPrice"
        >
          Promotional Price
        </label>
        <input
          id="promotionalPrice"
          min="0"
          name="promotionalPrice"
          step=".01"
          type="number"
        />
      </div>
      <button
        class="nimbus--button nimbus--button--primary"
        type="submit"
      >
        <i
          class="nimbus--button__icon nimbus--button__icon--start"
        >
          <svg
            aria-hidden="true"
            aria-label="Download"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: middle;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M2.5 14.75h11V16h-11zm.39-5.45l4.27 3.88a1.26 1.26 0 001.68 0l4.27-3.88a1.25 1.25 0 00-.84-2.18H9.8V1.25A1.25 1.25 0 008.55 0h-1.1A1.25 1.25 0 006.2 1.25v5.87H3.73a1.25 1.25 0 00-.84 2.18zm.84-.93h3.72V1.25h1.1v7.12h3.72L8 12.25 3.73 8.37z"
            />
          </svg>
        </i>
        Aceptar
      </button>
    </div>
  </form>
</div>
`;
